src/shared/particle_dynamics/fluid_dynamics/viscous_dynamics.hpp

```diff
@@ -18,143 +18,169 @@ ViscousForce<DataDelegationType>::ViscousForce(BaseRelationType &base_relation)
       viscous_force_(*this->particles_->template registerSharedVariable<Vecd>("ViscousForce")),
       smoothing_length_(this->sph_body_.sph_adaptation_->ReferenceSmoothingLength()) {}
 //=================================================================================================//
-template <class ViscosityType>
-ViscousForce<Inner<>, ViscosityType>::ViscousForce(BaseInnerRelation &inner_relation)
+template <typename ViscosityType, class KernelCorrectionType>
+ViscousForce<Inner<>, ViscosityType, KernelCorrectionType>::
+    ViscousForce(BaseInnerRelation &inner_relation)
     : ViscousForce<DataDelegateInner>(inner_relation),
-      ForcePrior(particles_, "ViscousForce"), mu_(particles_)
+      ForcePrior(particles_, "ViscousForce"), mu_(particles_),
+      kernel_correction_(particles_)
 {
     static_assert(std::is_base_of<ParticleAverage, ViscosityType>::value,
                   "ParticleAverage is not the base of ViscosityType!");
 }
 //=================================================================================================//
-template <class ViscosityType>
-void ViscousForce<Inner<>, ViscosityType>::interaction(size_t index_i, Real dt)
+template <typename ViscosityType, class KernelCorrectionType>
+void ViscousForce<Inner<>, ViscosityType, KernelCorrectionType>::interaction(size_t index_i, Real dt)
 {
     Vecd force = Vecd::Zero();
     const Neighborhood &inner_neighborhood = inner_configuration_[index_i];
     for (size_t n = 0; n != inner_neighborhood.current_size_; ++n)
     {
         size_t index_j = inner_neighborhood.j_[n];
+        const Vecd &e_ij = inner_neighborhood.e_ij_[n];
 
         // viscous force
         Vecd vel_derivative = (vel_[index_i] - vel_[index_j]) /
                               (inner_neighborhood.r_ij_[n] + 0.01 * smoothing_length_);
-        force += 2.0 * mu_(index_i, index_j) * vel_derivative *
+        force += e_ij.dot((kernel_correction_(index_i) + kernel_correction_(index_j)) * e_ij) *
+                 mu_(index_i, index_j) * vel_derivative *
```

Comment:
Here, we still use the conservative form (Bi+Bj)/2.

Reply:
Yes.